language/stdlib/modules/compare.move

```diff
@@ -0,0 +1,99 @@
+// `Compare::cmp<T>(t1, t2) == LESS_THEN` iff `t1 < t2` and (similarly)
+// `Compare::cmp<T>(t1, t2) == GREATER_THEN` iff `t1 > t2`.
+//
+// For all other types, the order is whatever the LCS encoding of the type and the comparison
+// strategy above gives you. One case where the order might be surprising is the `address` type.
+// Addresses are 16 byte hex values that LCS encodes with the identity function. The right to left,
+// byte-by-byte comparison means that (for example)
+// `Compare::cmp(0x01, 0x10)` == LESS_THAN` (as you'd expect), but
+// `Compare::cmp(0x100, 0x001) == LESS_THAN` (as you probably wouldn't expect). Keep this in mind
+// when writing code that compares addresses.
+//
+// Finally, a note on the constants `EQUAL`, `LESS_THAN`, and `GREATER_THAN`. Move does not have
+// signed integers, so we cannot use the usual 0, -1, 1 convention for this. Instead, we define a
+// new convention using u8 constants:
+// const EQUAL = 0u8
+// const LESS_THAN = 1u8
+// const GREATER_THAN = 2u8
+
+address 0x0:
+module Compare {
+    use 0x0::LCS;
+    use 0x0::Vector;
+
+    // Polymorphic comparison based on the LCS encoding of `v1` and `v2`. The lengths of the
+    // LCS-encoded vector are compared first, then the vectors are compared byte-by-byte from right
+    // to left if the lengths are equal.
+    // Returns either `EQUAL` (0u8), `LESS_THAN` (1u8), or `GREATER_THAN` (2u8).
+    public fun cmp<Value>(v1: &Value, v2: &Value): u8 {
+        cmp_vector_u8(&LCS::to_bytes(v1), &LCS::to_bytes(v2))
+    }
+
+    // Compare `v1` and `v2` using (1) the lengths of the vectors and (2) byte-by-byte comparison
+    // from right to left.
+    // Returns either `EQUAL` (0u8), `LESS_THAN` (1u8), or `GREATER_THAN` (2u8).
+    fun cmp_vector_u8(v1: &vector<u8>, v2: &vector<u8>): u8 {
+        let len1 = Vector::length(v1);
+        let len2 = Vector::length(v2);
+        let len_cmp = cmp_u64(len1, len2);
+        if (len_cmp != 0) {
+            // lengths unequal
+            return len_cmp
+        };
+        // else, lengths equal; must compare element-by-element
+
+        // LCS uses little endian encoding for all integer types, so we choose to compare from left
+        // to right. Going right to left would make the behavior of Compare.cmp diverge from the
+        // bytecode operators < and > on integer values (which would be bad).
+        let i = len1;
+        while (i > 0) {
+            i = i - 1;
+            let elem_cmp = cmp_u8(*Vector::borrow(v1, i), *Vector::borrow(v2, i));
+            if (elem_cmp != 0) {
+                // elements unequal; return result
+                return elem_cmp
+            };
+            // else, compare next element
+        };
+        // all elements equal
+        0u8
+    }
+
+    // Compare two `u8`'s
+    fun cmp_u8(i1: u8, i2: u8): u8 {
+        if (i1 == i2) {
+            0u8
+        } else if (i1 < i2) {
+            1u8
+        } else {
+            2u8
+        }
```

Comment:
Two things here. You can write `0` instead of `0u8` if you want. It will infer the correct type.

Similar to above about return. I like the look of the following (But maybe I have spent too much time in functional land):
```
if (i1 == i2) 0
else if (i1 < i2) 1
else 2
```
Note I haven't tried this and don't know if it parsers, but I think it will